Update VAT on Payment - more configurable - bank statement support

feketemihai opened the pull request:
Update vat on payment module, add configuration flag for move to shadow move lines, add configuration flag for using the same account, journal if not configured the related one, add ank statement reconciliation with vat onpayment.

feketemihai commented:
@eLBati @tafaRU Can you have a look on this, do you think it's ok for, i put configuration flag with default value on what was already in the module, add tests if i don't use the move_payment_lines_to_shadow_entry, and added another configuration flag if we don't have a set value on account, journal to use the same account...

feketemihai commented:
I still need to add tests for payments on bank statements.

eLBati commented on account_vat_on_payment/README.rst:
```diff
@@ -0,0 +1,74 @@
+
+By default you have to configure on tax code object, Related tax code used for real registrations on a VAT on payment basis.
+
+Depending on the selection field from the configuration, you can configure:
```
Please add an empty line after this

eLBati commented on account_vat_on_payment/README.rst:
```diff
@@ -0,0 +1,74 @@
+
+* Lorenzo Battistini <lorenzo.battistini@agilebg.com>
+* Alex Comba <alex.comba@agilebg.com>
+* Leonardo Pistone <lpistone@agmail.com>
```
You may want to add your name here

eLBati commented on account_vat_on_payment/account_bank_statement.py:
```diff
@@ -0,0 +1,299 @@
+#
+#    OpenERP, Open Source Management Solution
+#    Copyright (C) 2011-2012 Domsense s.r.l. (<http://www.domsense.com>).
+#    Copyright (C) 2014 Agile Business Group sagl (<http://www.agilebg.com>)
```
You may want to write your name here. Also, please use short headers, like https://github.com/OCA/maintainer-tools/blob/master/template/module/models/model_name.py

eLBati commented on account_vat_on_payment/__openerp__.py:
```diff
@@ -73,6 +55,16 @@
+        'test/account_invoice_8_real.yml',
+        'test/account_invoice_1_real_same_account.yml',
     ],
+    'images': [],
```
You may want to add screenshots. Otherwise, please remove empty item

feketemihai replied:
It seems that the new tests on travis check if the images item is declared.

tafaRU replied:
It should be raised only a warning, see https://github.com/OCA/maintainer-quality-tools/pull/298#issuecomment-159457373 as reference.

eLBati commented:
@feketemihai we are going to review and test the changes. Meanwhile, thanks for the big contribution!

eLBati commented on account_vat_on_payment/__openerp__.py:
```diff
@@ -73,6 +55,16 @@
+        'test/account_invoice_6_real.yml',
+        'test/account_invoice_7_real.yml',
+        'test/account_invoice_8_real.yml',
+        'test/account_invoice_1_real_same_account.yml',
```
It seems the `*_bank.yml` files are missing

feketemihai replied:
I will add them on monday.

eLBati replied:
Report: just tried with http://pastebin.com/K4JEJq6H and got

```
odoo/addons/account/account.py", line 1356, in button_validate
    _('You cannot validate this journal entry because account "%s" does not belong to chart of accounts "%s".') % (account.name, top_common.name))
except_orm: ('Error!', u'You cannot validate this journal entry because account "Bank Current Account - (test)" does not belong to chart of accounts "shadow COA".')
```

feketemihai replied:
Yes..that's why i want to check the bank test..i just copied the file to have the same invoices, even if i will change the id and the id of the partner, but i had the same issue with the tests that was already in the module, that's why i added the specific Bank Journal in vouchers...otherwise it didn't get the bank account move line.

feketemihai commented:
@eLBati Do you think it's worth it to add another config, of what we talked before the update to 8.0, against partner field on vat_on_payment, and to support directly the behaviour from this field, not to create a new fiscal position for it. At least in Romania is a fiscal attribute, so the supplier applies VAT on payment, he should note it on the invoice, and the client (even he doesn't apply VAT on Payment) shoudl register the invoice with this behaviour.

eLBati commented:
@feketemihai what is the problem with using fiscal positions?

eLBati commented:
@feketemihai I am testing the changes.
I configured the module following http://planet.agilebg.com/en/2012/10/vat-on-payment-treatment-with-openerp/
I created a new 'vat on payment' invoice.
I created a bank statement
While running 'reconcile' wizard I get

```
account_vat_on_payment/account_bank_statement.py", line 209, in _create_vat_on_payment_move
    mv_line_dict['counterpart_move_line_id'],
KeyError: 'counterpart_move_line_id'
```

Does this happen to you?

feketemihai commented:
@elBati It seems that i forgot to push the latest...i will do it on monday, with modified tests.

feketemihai commented:
@eLBati Tested on runbot and i didn't had any issue with the counterpart_move_line_id, could you check it out, on runbot i configure the romanian chart of accounts, taxes, added shadow account for 707 (incomes), added shadow tax codes for general (24In, 24Out) tax codes, added shadow for bank, cash journal...made partial payments in bank statement, cash register, pay all in bank statement and no error...i will make the tests now. Invoice is on Axelor.

feketemihai commented:
@eLBati @tafaRU Any other comments on this?

eLBati commented:
Hello @feketemihai , I didn't notice you added tests.
I will check this in the next days

feketemihai commented:
@eLBati Thanks.

eLBati commented on account_vat_on_payment/test/account_invoice_1_bank.yml:
```diff
@@ -0,0 +1,143 @@
+-
+  I create shadow Tax Code
```
Why not using the same tax code configuration created in `account_invoice_1.yml`?

eLBati commented on account_vat_on_payment/test/account_invoice_1_bank.yml:
```diff
@@ -0,0 +1,143 @@
+    sign: 1
+    vat_on_payment_related_tax_code_id: shadow_tax_code
+-
+  I create Tax 10%
```
`tax10` and `tax20` are already created by `account_invoice_1.yml`
I think you can use them

feketemihai commented:
@eLBati Sorry, i removed the declaration, i tested only the bank statements moves so i moved the configuration...but it's ok i removed them to stay only at the first test.

eLBati commented on account_vat_on_payment/account_bank_statement.py:
```diff
@@ -0,0 +1,289 @@
+                    })
+        return True
+
+    def _create_vat_on_payment_move(self, cr, uid, bank_line, mv_line_dicts,
```
Do you think it's possible to extract 1 or more methods from this one?
For example, a method to compute 'debit_at_old_rate' and 'credit_at_old_rate', another for 'new_line_amount'.

feketemihai replied:
@eLBati I will do it today, but i think it's more usefull to do it directly on the odoo base, so we can use afterwards...

feketemihai replied:
@eLBati PR made on Odoo in https://github.com/odoo/odoo/pull/10447

eLBati commented on account_vat_on_payment/account_bank_statement.py:
```diff
@@ -0,0 +1,289 @@
+                if bank_line.company_id.vat_payment_lines == 'shadow_move':
+                    line_to_create['move_id'] = shadow_move_id
+                else:
+                    line_to_create['move_id'] = shadow_move_id
```
Here `shadow_move_id` is always used

feketemihai commented:
I left it this way because the real move is created in the reconcile method, so to have the same behaviour as vouchers i will have to cancel the move_id, check by accounts which one corresponds to the invoice line, move to shadow move id an the vat lines from shadow move id to move it to the real one...it's not that easy inheritable the method...so i was quite happy with all the vat on payment moves on the shadow move.

eLBati commented on account_vat_on_payment/account_invoice.py:
```diff
@@ -35,18 +18,22 @@ def _set_vat_on_payment_account(self, cr, uid, line_tuple, context=None):
+            line_tuple[2]['account_id'] = real_account
         return line_tuple
 
     def _set_vat_on_payment_tax_code(self, cr, uid, line_tuple, context=None):
```
Why not also checking `vat_config_error` in this method?

feketemihai replied:
@eLBati I don't check it there because for the taxes it's mandatory to have the shadow tax, you can use the same account or journal, but for taxes, this is the main objective of the module, to move the tax to an "not elligible" one untill the payment.

eLBati replied:
Ok thanks

eLBati commented:
@feketemihai Hi.
In PR https://github.com/feketemihai/account-payment/pull/1 you can see various improvements and a refactoring of the bank statement part.

I did these changes because I got into errors like 

```
account_vat_on_payment/account_bank_statement.py", line 185, in _create_vat_on_payment_move
mv_line_dict['counterpart_move_line_id'],
KeyError: 'counterpart_move_line_id'
```

and I found some part of the implementation a bit confusing.

Mainly, I followed the same approach used in vouchers, using `allocated_amounts_grouped_by_invoice` method.

Please review https://github.com/feketemihai/account-payment/pull/1 and tell me if you have any doubts.

Thanks

feketemihai commented:
@eLBati Thanks for the help and time to review and work on the bank statements payments with vat  on payment.

eLBati commented:
@feketemihai please rebase to fix conflicts introduced after https://github.com/OCA/account-payment/pull/64

Tell me if you have any trouble

feketemihai commented:
@eLBati Done...everything is ok now.

eLBati commented:
Thanks a lot :+1:

feketemihai commented:
@OCA/accounting-maintainers Can i have a last review, please?

feketemihai commented:
@OCA/accounting-maintainers Can someone merge this if it's ok?

tafaRU commented:
:+1:

eLBati commented:
Thanks @feketemihai

feketemihai commented:
@eLBati Thank you for accepting it and for time reviewing.
